sidecar/proxy/adapter.go

```diff
@@ -14,14 +14,8 @@
 
 package proxy
 
-import (
-	"testing"
-
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
-)
-
-func TestPackage(t *testing.T) {
-	RegisterFailHandler(Fail)
-	RunSpecs(t, "Poll Suite")
+// Adapter interface for proxy implementations.
+type Adapter interface {
```

Comment:
Have you considered extending the `Adapter` so that it embeds both `DiscoveryListener` and `RulesListener`?
```golang
type Adapter interface {
	monitor.DiscoveryListener
	monitor.RulesListener

 	Start() error
 	Stop() error
}
```

We anyway expect all proxy adapters to be somehow able to react to discovery/rules changes. This also decouples the various proxy adapters from `DiscoveryMonitor` and `RulesMonitor`. And as a bonus, we infuse some more meaning to the proxy adapter abstraction, other than "something that can be Start()'ed and Stop()'ed".

These are not _major_ benefits, but maybe somewhat cleaner design.